### PR TITLE
otelgrpc: Refine tests to use a net socket instead of a buffer

### PR DESCRIPTION
**Pull request by MadVikingGod**
These are the changes proposed by @pellared in https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4322.

Currently, I haven't addressed the metric failure because when #4322 lands, those attributes will be removed.

### CODE REVIEW

File: instrumentation/google.golang.org/grpc/otelgrpc/test/grpc_test.go [hunk 1]
```diff
@@ -411,7 +437,10 @@ func checkStreamServerSpans(t *testing.T, spans []trace.ReadOnlySpan) {
 		semconv.RPCService("grpc.testing.TestService"),
 		otelgrpc.RPCSystemGRPC,
 		otelgrpc.GRPCStatusCodeKey.Int64(int64(codes.OK)),
-	}, streamInput.Attributes())
+		semconv.NetSockPeerAddr("127.0.0.1"),
+	}, filterKVs(streamInput.Attributes(), semconv.NetSockPeerPortKey))
+	// Client port is random
+	assert.True(t, existKVs(streamInput.Attributes(), semconv.NetSockPeerAddrKey))
```
Comment (pellared) [inline instrumentation/google.golang.org/grpc/otelgrpc/test/grpc_test.go:443]: How about creating a helper like

```go
func findAttribute(attr []attribute.KeyValue, key attribute.Key) attribute.KeyValue
	for _, a := range attr {
		if a.Key == key {
			return a
		}
	}
	return attribute.KeyValue{}
}
```

and use it like this:

```go
netSockPeerPort := findAttribute(streamInput.Attributes(), semconv.NetSockPeerPortKey)
require.NoZero(t, netSockPeerPort)
assert.ElementsMatch(t, []attribute.KeyValue{
		semconv.RPCMethod("StreamingInputCall"),
		semconv.RPCService("grpc.testing.TestService"),
		otelgrpc.RPCSystemGRPC,
		otelgrpc.GRPCStatusCodeKey.Int64(int64(codes.OK)),
		semconv.NetSockPeerAddr("127.0.0.1"),
		netSockPeerPort 
	}, streamInput.Attributes())
```

Comment (pellared) [reply instrumentation/google.golang.org/grpc/otelgrpc/test/grpc_test.go:443]: Even this should work

```go
ssert.ElementsMatch(t, []attribute.KeyValue{
		semconv.RPCMethod("StreamingInputCall"),
		semconv.RPCService("grpc.testing.TestService"),
		otelgrpc.RPCSystemGRPC,
		otelgrpc.GRPCStatusCodeKey.Int64(int64(codes.OK)),
		semconv.NetSockPeerAddr("127.0.0.1"),
		findAttribute(streamInput.Attributes(), semconv.NetSockPeerPortKey),
	}, streamInput.Attributes())
```

Comment (MadVikingGod) [reply instrumentation/google.golang.org/grpc/otelgrpc/test/grpc_test.go:443]: I don't think we would want to rely on the empty attribute.  There is a chance both the port attribute doesn't exist, returning the empty attribute, and the instrumentation puts an empty attribute in the list.  Together this test would pass.

If the call was `findAttribute() (attribute.KeyValue, bool)` that would work, but doesn't really simplify the tests.

Comment (pellared) [reply instrumentation/google.golang.org/grpc/otelgrpc/test/grpc_test.go:443]: Then the first code snippet: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4503#discussion_r1380830445 would work for you thanks to 

```go
require.NoZero(t, netSockPeerPort)
```

Personally, I think that a case when the instrumentation puts an empty attribute in the list is such an edge case that I would not worry about it.

**Comment by pellared** [comment]
I am fine with the tests in this PR and I would love to see them as a part of https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4322

As for me https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4503#discussion_r1380830445 can be addressed in a separate PR.

**Comment by MrAlias** [comment]
Given this is less than 300 lines and we've talked a decent amount about this already, I'm in favor of combining this with the PR this is intended to test. 👍

File: instrumentation/google.golang.org/grpc/otelgrpc/test/grpc_test.go [hunk 2]
```diff
@@ -47,42 +48,47 @@ var wantInstrumentationScope = instrumentation.Scope{
...
 const bufSize = 2048
 
-func doCalls(cOpt []grpc.DialOption, sOpt []grpc.ServerOption) error {
-	l := bufconn.Listen(bufSize)
-	defer l.Close()
-
-	s := grpc.NewServer(sOpt...)
-	pb.RegisterTestServiceServer(s, interop.NewTestServer())
+// newGrpcTest creats a grpc server, starts it, and executes all the calls, closes everything down.
+func newGrpcTest(listener net.Listener, cOpt []grpc.DialOption, sOpt []grpc.ServerOption) (err error) {
```
Comment (pellared) [inline instrumentation/google.golang.org/grpc/otelgrpc/test/grpc_test.go:52]:
```suggestion
func newGrpcTest(listener net.Listener, cOpt []grpc.DialOption, sOpt []grpc.ServerOption) error {
```

**Review by pellared — APPROVED** [review]

**Review by dashpole — APPROVED** [review]

File: instrumentation/google.golang.org/grpc/otelgrpc/test/grpc_stats_handler_test.go [hunk 3]
```diff
@@ -46,14 +47,18 @@ func TestStatsHandler(t *testing.T) {
 	serverMetricReader := metric.NewManualReader()
 	serverMP := metric.NewMeterProvider(metric.WithReader(serverMetricReader))
 
-	assert.NoError(t, doCalls(
+	listener := bufconn.Listen(bufSize)
+	defer listener.Close()
```
Comment (pellared) [inline instrumentation/google.golang.org/grpc/otelgrpc/test/grpc_stats_handler_test.go:51]: I guess that as a follow-up PR we would like to use `net.Listen("tcp", "127.0.0.1:0")` here instead.

Comment (pellared) [reply instrumentation/google.golang.org/grpc/otelgrpc/test/grpc_stats_handler_test.go:51]: https://github.com/open-telemetry/opentelemetry-go-contrib/issues/4532

**Comment by pellared** [comment]
This https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4503/commits/be618e6fbac4d63290f69f23e957480317b4b123 attempts to fix https://github.com/open-telemetry/opentelemetry-go-contrib/issues/1352 and https://github.com/open-telemetry/opentelemetry-go-contrib/issues/4416
